src/contracts/rewardsV2/VaultSnapshotRewards.sol

```diff
@@ -0,0 +1,422 @@
+        return _vaultSnapshotRewardsStorage()._rewards[vault][network][token][index];
+    }
+
+    /**
+     * @inheritdoc IVaultSnapshotRewards
+     */
+    function lastUnclaimedReward(
+        address account,
+        address vault,
+        address network,
+        address token
+    ) public view returns (uint256) {
+        return _vaultSnapshotRewardsStorage()._lastUnclaimedReward[account][vault][network][token];
+    }
+
+    /**
+     * @inheritdoc IVaultSnapshotRewards
+     */
+    function distributeVaultSnapshotRewards(
+        bytes32 subnetwork,
+        address token,
+        address vault,
+        uint256 amount,
+        uint48 timestamp,
+        bytes calldata activeSharesHint,
+        bytes calldata activeStakeHint
+    ) public {
+        // Check authorization - either network middleware or network itself
+        address network = subnetwork.network();
+        if (
+            network != msg.sender
+                && INetworkMiddlewareService(NETWORK_MIDDLEWARE_SERVICE).middleware(network) != msg.sender
+        ) {
+            revert NotNetworkOrMiddleware();
+        }
+
+        // Validate vault
+        if (!IRegistry(VAULT_FACTORY).isEntity(vault)) {
+            revert InvalidVault();
+        }
+
+        // Validate timestamp
+        if (timestamp >= block.timestamp) {
+            revert InvalidRewardTimestamp();
+        }
+
+        // Cache active shares if not already cached
+        if (_vaultSnapshotRewardsStorage()._activeSharesCache[vault][timestamp] == 0) {
+            uint256 activeShares = IVault(vault).activeSharesAt(timestamp, activeSharesHint);
+            uint256 activeStake = IVault(vault).activeStakeAt(timestamp, activeStakeHint);
+
+            if (activeShares == 0 || activeStake == 0) {
+                revert InvalidRewardTimestamp();
+            }
+
+            _vaultSnapshotRewardsStorage()._activeSharesCache[vault][timestamp] = activeShares;
+        }
+
+        // Transfer tokens
+        uint256 balanceBefore = IERC20(token).balanceOf(address(this));
+        IERC20(token).safeTransferFrom(msg.sender, address(this), amount);
+        amount = IERC20(token).balanceOf(address(this)) - balanceBefore;
+
+        if (amount == 0) {
+            revert InsufficientReward();
+        }
+
+        uint256 maxFee = FeeRegistry(FEE_REGISTRY).MAX_FEE();
+        // Get curator fee from FeeRegistry
+        uint256 curatorFee = IFeeRegistry(FEE_REGISTRY).getCuratorFee(vault, network);
+        uint256 curatorFeeAmount = amount.mulDiv(curatorFee, maxFee); // Assuming 10000 as base
+
+        // Get operators fee from FeeRegistry
+        uint256 operatorsFee = IFeeRegistry(FEE_REGISTRY).getOperatorsFee(vault, network);
+        uint256 operatorsFeeAmount = amount.mulDiv(operatorsFee, maxFee);
+
+        // Deduct protocol fees from the remaining amount
+        uint256 protocolFees = _deductProtocolFees(REWARDS_TYPE_VAULT_SNAPSHOT, network, token, amount);
```

Comment:
Protocol fees are deducted from the whole amount, and curator/operators from the remaining part